README.md
# glmnettools

<!-- badges: start -->
[![Project Status: WIP – Initial development is in progress, but there has not yet been a stable, usable release suitable for the public.](https://www.repostatus.org/badges/latest/wip.svg)](https://www.repostatus.org/#wip)
[![license](http://img.shields.io/badge/license-GPL%20%28%3E=%203%29-brightgreen.svg?style=flat)](http://www.gnu.org/licenses/gpl-3.0.html)
[![R build status](https://github.com/ampel-leipzig/glmnettools/workflows/R-CMD-check/badge.svg)](https://github.com/ampel-leipzig/glmnettools/actions)
[![Codecov test
coverage](https://codecov.io/gh/ampel-leipzig/glmnettools/branch/main/graph/badge.svg)](https://codecov.io/gh/ampel-leipzig/glmnettools?branch=main)
<!-- badges: end -->

Additional Tools for [glmnet](https://cran.r-project.org/package=glmnet).

## Description

`glmnettools` provides repeated cross-validation for `glmnet`.

Basic usage: https://ampel-leipzig.github.io/glmnettools/articles/glmnettools.html

## Install

Currently this package is in development and just available via the github
repository.

To install the latest development version use:

```r
remotes::install_github("ampel-leipzig/glmnettools")
```

## Contact

You are welcome to:

* submit suggestions and bug-reports at: <https://github.com/ampel-leipzig/glmnettools/issues>
* send a pull request on: <https://github.com/ampel-leipzig/glmnettools/pulls>
* compose an e-mail to: <mail@sebastiangibb.de>


NEWS.md
# glmnettools 0.0

## Changes in 0.0.4

- Pass `time.inc` argument to `rms::cph` in `as.cph`.

## Changes in 0.0.3

- Add `as.coxph` and `as.cph`.

## Changes in 0.0.2

- Add `predict.rcv.glmnet`.
- Add manual page for `plot.rcv.glmnet`.

## Changes in 0.0.1

- Initial version.
- Add `rcv.glmnet` and `plot.rcv.glmnet`.


tests/testthat.R
library("testthat")
library("glmnettools")

test_check("glmnettools")


tests/testthat/test_utils.R
test_that(".s2numeric", {
    expect_error(.s2numeric("foo"), "cv.glmnet.* object")

    set.seed(1010)
    n <- 200
    p <- 30
    x <- matrix(rnorm(n * p), n, p)
    nzc <- trunc(p/10)
    beta <- rnorm(nzc)
    fx <- x[, seq(nzc)] %*% beta
    eps <- rnorm(n) * 5
    y <- drop(fx + eps)

    rcv <- rcv.glmnet(
        x, y, nrepcv = 1, nfolds = 3, mc.cores = 1, trace.it = FALSE
    )
    expect_error(.s2numeric(rcv, "foo"))
    expect_equal(.s2numeric(rcv, "lambda.min"), rcv[["lambda.min"]])
    expect_equal(.s2numeric(rcv, c(0.1, 0.2)), c(0.1, 0.2))
})


tests/testthat/test_rcv.glmnet.R
test_that("rcv.glmnet", {
    # taken from glmnet::cv.glmnet
    set.seed(1010)
    n <- 1000
    p <- 100
    nzc <- trunc(p/10)
    x <- matrix(rnorm(n * p), n, p)
    beta <- rnorm(nzc)
    fx <- x[, seq(nzc)] %*% beta
    eps <- rnorm(n) * 5
    y <- drop(fx + eps)
    set.seed(1011)
    cv <- list(
        cv1 = cv.glmnet(x, y, nfolds = 3),
        cv2 = cv.glmnet(x, y, nfolds = 3)
    )
    set.seed(1011)
    rcv <- rcv.glmnet(
        x, y, nrepcv = 2, nfolds = 3, mc.cores = 1, trace.it = FALSE
    )
    expect_equal(
        rcv$cvm,
        colMeans(do.call(rbind, list(cv[[1]]$cvm, cv[[2]]$cvm)))
    )
    expect_equal(
        rcv$index,
        matrix(
            c(25, 17), nrow = 2,
            dimnames = list(c("min", "1se"), c("Lambda"))
        )
    )

#' # each base::graphics plot function must be wrapped by an anonymous function
#' # that could be called by `vdiffr::expect_doppelganger()`
#' # use `vdiffr::manage_cases()` to add new/verify changed plots
#' vdiffr::manage_cases()
    vdiffr::expect_doppelganger("rcv.glmnet-plot", function()plot(rcv))
})

test_that("predict.rcv.glmnet", {
    # taken from glmnet::cv.glmnet
    set.seed(10101)
    n <- 500
    p <- 30
    nzc <- trunc(p / 10)
    x <- matrix(rnorm(n * p), n, p)
    beta <- rnorm(nzc)
    fx <- x[, seq(nzc)] %*% beta / 3
    hx <- exp(fx)
    ty <- rexp(n, hx)
    tcens <- rbinom(n = n, prob = 0.3, size = 1)  # censoring indicator
    y <- cbind(time = ty, status = 1 - tcens)

    expect_error(
            predict(
                rcv.glmnet(
                    x, y[, "status"], nrepcv = 1, nfolds = 3,
                    mc.cores = 1, trace.it = FALSE
                ),
                type = "survival"
            ),
            "family = .*cox.*"
    )

    rcvobj <- rcv.glmnet(
        x, y, family = "cox", nrepcv = 1, nfolds = 3,
        mc.cores = 1, trace.it = FALSE
    )
    expect_error(
        predict(rcvobj, type = "survival", s = c(0.1, 0.2)), "length 1"
    )

    expect_equal(
        predict(rcvobj, newx = x[1:5,], type = "link", s = "lambda.1se"),
        predict(
            rcvobj$glmnet.fit, newx = x[1:5,],
            type = "link", s = rcvobj[["lambda.1se"]]
        )
    )

    skip_if_not_installed("survival")
    requireNamespace("survival")

    expect_equal(
        predict(
            rcvobj,
            x = x, y = survival::Surv(y[, "time"], y[, "status"]),
            newx = x[1:5,],
            type = "survival", s = "lambda.1se", times = c(0, 7)
        ),
        matrix(
            c(rep(1, 5), 0.1312533, 0.05047742, 0.002480262, 0.01006882,
              0.02348179),
            byrow = TRUE, nrow = 2, ncol = 5,
            dimnames = list(NULL, c(as.character(1:5)))
        )
    )

    expect_equal(
        dim(predict(
            rcvobj,
            x = x, y = survival::Surv(y[, "time"], y[, "status"]),
            newx = x[1:5,],
            type = "survival", s = "lambda.1se"
        )), c(343, 5)
    )
})


R/as.R
#' Convert coxnet objects
#'
#' These functions convert `coxnet` objects into popular survival objects like
#' `survival::coxph` or `rms::cph`.
#'
#' @param object object that should be converted.
#'
#' @rdname as
#' @author Sebastian Gibb
#' @export
as.coxph <- function(object, ...) {
    UseMethod("as.coxph")
}

#' @rdname as
#' @param s `character`/`numeric`, value(s) of the penality parameter `lambda`.
#' @param x `matrix`, as in `glmnet`.
#' @param y response as in `glmnet`.
#' @param \dots further arguments passed to the underlying function
#' (`survival::coxph` or `rms::cph`).
#' @seealso [`glmnet::glmnet()`]
#' @method as.coxph rcv.glmnet
#' @export
as.coxph.rcv.glmnet <- function(object, s, x, y, ...) {
    as.coxph(object$glmnet.fit, s = .s2numeric(object, s), x = x, y = y, ...)
}

#' @rdname as
#' @method as.coxph coxnet
#' @export
as.coxph.coxnet <- function(object, s, x, y, ...) {
    requireNamespace("survival")
    if (!is(y, "Surv") && is(y, "matrix") && ncol(y) == 2L)
        y <- survival::Surv(y[, 1L], y[, 2L])
    glmnet:::mycoxph(object = object, s = s, x = x, y = y, ...)
}

#' @rdname as
#' @export
as.cph <- function(object, ...) {
    UseMethod("as.cph")
}

#' @rdname as
#' @seealso [`rms::cph()`]
#' @method as.cph rcv.glmnet
#' @export
as.cph.rcv.glmnet <- function(object, s, x, y, ...) {
    as.cph(object$glmnet.fit, s = .s2numeric(object, s), x = x, y = y, ...)
}

#' @rdname as
#' @importFrom stats as.formula predict
#' @param time.inc `numeric(1)`, time increment, see [`rms::cph()`] for details.
#' @method as.cph coxnet
#' @export
as.cph.coxnet <- function(object, s, x, y, time.inc = 30, ...) {
    requireNamespace("rms")
    requireNamespace("survival")
    if (!is(y, "Surv") && is(y, "matrix") && ncol(y) == 2L)
        y <- survival::Surv(y[, 1L], y[, 2L])

    ## taken from glmnet version 4.1-1 glmnet:::mycoxph
    args <- list(...)
    glmnet_call_names <- names(object$call)[-1]
    args$object <- object
    args$newx <- x
    args$s <- s
    args$x <- x
    args$y <- y
    if ("offset" %in% glmnet_call_names) {
        args$newoffset <- rep(0, length.out = nrow(x))
    }
    eta <- do.call(predict, args)
    coxphargs <- list(
        formula = "y ~ X1", data = data.frame(y, eta),
        init = 1, iter = 0, x = TRUE, y = TRUE, surv = TRUE,
        time.inc = time.inc
    )
    if ("strata" %in% names(attributes(y))) {
        coxphargs$data$strata <- attr(y, "strata")
        coxphargs$formula <- paste(coxphargs$formula, "+ strata(strata)")
    }
    if ("weights" %in% glmnet_call_names) {
        coxphargs$weights <- args$weights
    }
    if ("offset" %in% glmnet_call_names) {
        coxphargs$data$offset <- args$offset
        coxphargs$formula <- paste(coxphargs$formula, "+ offset(offset)")
    }
    coxphargs$formula <- as.formula(coxphargs$formula)
    do.call(rms::cph, coxphargs)
}


R/rcv.glmnet.R
#' Repeated Cross-Validation for cv.glmnet
#'
#' This functions returns the best lambda of repeated [`glmnet::cv.glmnet()`]
#' calls.
#'
#' @param x `matrix`, as in `cv.glmnet`.
#' @param y response as in `cv.glmnet`.
#' @param lambda `numeric`, optional user-supplied lambda sequence; default is
#' `NULL` and `glmnet` chooeses its own sequence.
#' @param nrepcv `integer(1)`, number of repeated cross-validations (outer
#' loop).
#' @param nfolds `integer`, number of folds, same as in `cv.glmnet`.
#' @param \dots further arguments passed to `cv.glmnet`.
#' @param trace.it `integer`, if `trace.it = 1`, then a progress bar is
#' displayed.
#' @param mc.cores `integer`, number of cores used for parallelisation.
#'
#' @return An object of class `rcv.glmnet` that extends the `cv.glmnet`
#' class.
#'
#' @author Sebastian Gibb
#' @seealso [`glmnet::cv.glmnet()`]
#' @references
#' Jerome Friedman, Trevor Hastie, Robert Tibshirani (2010).
#'  Regularization Paths for Generalized Linear Models via Coordinate
#'  Descent. Journal of Statistical Software, 33(1), 1-22. URL
#'  \url{https://www.jstatsoft.org/v33/i01/}.
#'
#'  Noah Simon, Jerome Friedman, Trevor Hastie, Rob Tibshirani (2011).
#'  Regularization Paths for Cox's Proportional Hazards Model via
#'  Coordinate Descent. Journal of Statistical Software, 39(5), 1-13. URL
#'  \url{https://www.jstatsoft.org/v39/i05/}.
#'
#' @importFrom parallel mclapply
#' @importFrom glmnet cv.glmnet
#' @usage rcv.glmnet(
#'     x, y,
#'     lambda = NULL,
#'     nrepcv = 100L, nfolds = 10L,
#'     ...,
#'     trace.it = interactive(), mc.cores = getOption("mc.cores", 1L)
#' )
#' @export rcv.glmnet
#' @examples
#' # Examples taken from ?"glmnet::cv.glmnet"
#' set.seed(1010)
#' n <- 1000
#' p <- 100
#' nzc <- trunc(p/10)
#' x <- matrix(rnorm(n * p), n, p)
#' beta <- rnorm(nzc)
#' fx <- x[, seq(nzc)] %*% beta
#' eps <- rnorm(n) * 5
#' y <- drop(fx + eps)
#' set.seed(1011)
#' # nrepcv should usually be higher but to keep the runtime of the example low
#' # we choose 2 here
#' rcvob <- rcv.glmnet(x, y, nrepcv = 2, nfolds = 3)
#' plot(rcvob)
#' title("Gaussian Family", line = 2.5)
#' coef(rcvob)
#' predict(rcvob, newx = x[1:5, ], s = "lambda.min")
rcv.glmnet <- function(x, y, lambda = NULL, nrepcv = 100L, nfolds = 10L, ...,
                          trace.it = interactive(),
                          mc.cores = getOption("mc.cores", 1L)) {
    if (as.logical(trace.it) && requireNamespace("pbapply")) {
        rcv <- pbapply::pblapply(
            integer(nrepcv),
            function(i)cv.glmnet(
                x = x, y = y, lambda = lambda, nfolds = nfolds, ...
            ),
            cl = mc.cores
        )
    } else {
        rcv <- parallel::mclapply(
            integer(nrepcv),
            function(i)cv.glmnet(
                x = x, y = y, lambda = lambda, nfolds = nfolds, ...
            ),
            mc.cores = mc.cores
        )
    }

    cvm <- cvsd <- matrix(
        NA_real_, nrow = length(rcv[[1L]]$lambda), ncol = nrepcv
    )

    for (i in seq(along=rcv)) {
        cvm[, i] <- rcv[[i]]$cvm
        cvsd[, i] <- rcv[[i]]$cvsd
    }

    cvm <- rowMeans(cvm)
    cvsd <- rowMeans(cvsd)
    i <- which.min(cvm)

    index <- matrix(
        c(i, which(cvm < (cvm[i] + cvsd[i]))[1L]),
        nrow = 2L, ncol = 1L, dimnames = list(c("min", "1se"), "Lambda")
    )

    out <- list(
        call = match.call(),
        cvm = cvm,
        cvsd = cvsd,
        cvup = cvm + cvsd,
        cvlo = cvm - cvsd,
        index = index,
        nrepcv = nrepcv,
        nfolds = nfolds,
        lambda.min = rcv[[1L]]$lambda[index["min",]],
        lambda.1se = rcv[[1L]]$lambda[index["1se",]]
    )
    out <- c(out, rcv[[1L]][c("glmnet.fit", "lambda", "name", "nzero")])
    class(out) <- c("rcv.glmnet", class(rcv[[1L]]))
    out
}

#' Plot the cross-validation curve
#'
#' This functions plots the aggregated cross-validation curve produced by
#' [`rcv.glmnet()`].
#'
#' @param x `rcv.glmnet` object.
#' @param \dots further arguments passed to `cv.glmnet`.
#'
#' @author Sebastian Gibb
#' @seealso [`glmnet::cv.glmnet()`]
#' @importFrom graphics title
#' @method plot rcv.glmnet
#' @export
plot.rcv.glmnet <- function(x, ...) {
    NextMethod()
    title(
        sub = paste(
            "Averaged across", x$nrepcv, "repeated cross-validations",
            "each with", x$nfolds, "folds."
        ),
        adj = 0L
    )
}

#' Predictions for a `rcv.glmnet` object
#'
#' Compute fitted values for a model fitted by `rcv.glmnet`.
#'
#' @param object `rcv.glmnet` object.
#' @param newx `matrix`, of new values for `x` at which predictions are to be
#' made.
#' @param s `character`/`numeric`, value(s) of the penality parameter `lambda`.
#' See [`glmnet::predict.cv.glmnet()`] for details.
#' @param type `character`, type of prediction. For `"survival"`
#' the predicted survival is returned for each timepoint in `times`. All other
#' types are passed to `predict.cv.glmnet`
#' See [`glmnet::predict.cv.glmnet()`] for details.
#' @param times `numeric`, vector of times, the returned matrix will contain one
#' row for each time. See [`survival::summary.survfit()`] for details.
#' @param \dots further arguments passed to `predict.cv.glmnet`.
#'
#' @return The object returned depends on the \dots arguments.
#' See [`glmnet::predict.cv.glmnet()`] for details. For `type = "survival"` the
#' returned value is a `matrix` with one row per `times` element and one column
#' for each row in `newx`.
#' @author Sebastian Gibb
#' @seealso [`rcv.glmnet()`], [`glmnet::predict.cv.glmnet()`],
#' [`survival::summary.survfit()`]
#' @importFrom methods is
#' @method predict rcv.glmnet
#' @export
#' @examples
#' # Example adapted from ?"glmnet::cv.glmnet"
#' set.seed(10101)
#' n <- 500
#' p <- 30
#' nzc <- trunc(p / 10)
#' x <- matrix(rnorm(n * p), n, p)
#' beta <- rnorm(nzc)
#' fx <- x[, seq(nzc)] %*% beta / 3
#' hx <- exp(fx)
#' ty <- rexp(n, hx)
#' tcens <- rbinom(n = n, prob = 0.3, size = 1)  # censoring indicator
#' # y <- Surv(ty, 1-tcens) with library("survival")
#' y <- cbind(time = ty, status = 1 - tcens)
#' # nrepcv should usually be higher but to keep the runtime of the example low
#' # we choose 2 here
#' rcvob <- rcv.glmnet(x, y, family = "cox", nrepcv = 2, nfolds = 3)
#' predict(
#'     rcvob,
#'     newx = x[1:5,], x = x, y = survival::Surv(y[, "time"], y[, "status"]),
#'     type = "survival", times = c(0, 7), s = "lambda.1se"
#' )
predict.rcv.glmnet <- function(object, newx, s = c("lambda.1se", "lambda.min"),
                               type = c("link", "response", "coefficients",
                                        "nonzero", "class", "survival"),
                               times, ...) {
    type <- match.arg(type)

    if (type == "survival") {
        if (!is(object$glmnet.fit, "coxnet"))
            stop(
                "Survival prediction is just supported for ",
                "'rcv.glmnet(..., family = \"cox\")."
            )
        requireNamespace("survival")
        s <- .s2numeric(object, s)
        if (length(s) != 1L)
            stop("'s' has to be an 'numeric' or 'character' of length 1.")
        args <- list(...)
        srvft <- survival::survfit(object, newx = newx, s = s, ...)
        summaryargs <-
            args[names(args) %in% c("censored", "scale", "extend", "rmean")]

        if (!missing(times))
            summaryargs <- append(summaryargs, list(times = times))
        do.call(
            summary,
            append(list(object = srvft), summaryargs)
        )$surv
    } else
        NextMethod()
}
